Arduino gpio fix

hakehuang opened the pull request:
st platforms require input and output on same port, add support for such constrains

#fixes: #78381

erwango commented:
> st platforms require input and output on same port, add support for such constrains

This is wrong.

`tests/drivers/gpio/gpio_basic_api/` expects test pin 1 and test pin 2 to be on the same port.

With [your PR](https://github.com/zephyrproject-rtos/zephyr/pull/77699) pins used for the test have been moved (from A0/A1 to D2/D3) and on most STM32, D2/D3 pins are not on the same GPIO bank, hence test contraint is no more respected.

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,20 +36,66 @@ tests:
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and dt_res_prop_addr_cmp("in-gpios","out-gpios", 0) and CONFIG_SOC_FAMILY_STM32
+    extra_args: "DTC_OVERLAY_FILE=st_arduino.overlay"
+    extra_configs:
+      - CONFIG_ONE_PORT=y
```
Default configuration, not required

hakehuang replied:
yes

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,20 +36,66 @@ tests:
+    extra_configs:
+      - CONFIG_ONE_PORT=y
+    integration_platforms:
+      - stm32mp157c_dk2
```
Any other one would be better (say disco_l475_iot1)

hakehuang replied:
ok

erwango replied:
Thanks

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,20 +36,66 @@ tests:
+      - stm32mp157c_dk2
+    platform_exclude:
+      # below boards are mikro bus not arduino
+      - da14695_dk_usb
```
Not STM32_SOC_FAMILY

hakehuang replied:
this board declare as arduino interface in <board?.yaml but actually mikro bus interface in dts, do you know who is the maintainer for this board?

erwango replied:
I don't know who is maintaining it but I think it would be justified if you change it.

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,20 +36,66 @@ tests:
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and not CONFIG_SOC_FAMILY_STM32
     extra_args: "DTC_OVERLAY_FILE=arduino.overlay"
+    extra_configs:
+      - CONFIG_ONE_PORT=n
```
It seems wrong that `drivers.gpio.2pin_arduino:` which looks as the default test is set with `CONFIG_ONE_PORT=n` which is not default test configuration

hakehuang commented:
> With [your PR](https://github.com/zephyrproject-rtos/zephyr/pull/77699) pins used for the test have been moved (from A0/A1 to D2/D3) and on most STM32, D2/D3 pins are not on the same GPIO bank, hence test contraint is no more respected.

@erwango , thanks for the confirmation, I am modify the cases to split the port dependency.

hakehuang commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,20 +36,66 @@ tests:
       - gpio
       - arduino_gpio
-    filter: dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+    filter: >
+      dt_compat_enabled("test-gpio-basic-api") and dt_compat_enabled("arduino-header-r3")
+      and not CONFIG_SOC_FAMILY_STM32
```
ok, I see, it is not IP limitation, just case limitation, let me update

erwango replied:
So, if the test doesn't require anymore to have both pins on same GPIO bank, this "STM32 specific case" is not required anymore.

erwango commented on tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c:
```diff
@@ -1,21 +1,22 @@
 #define ALL_BITS ((gpio_port_value_t)-1)
 
-static const struct device *const dev = DEVICE_DT_GET(DEV);
+static const struct device *const dev_in = DEVICE_DT_GET(DEV_IN);
+static const struct device *const dev_out = DEVICE_DT_GET(DEV_OUT);
```
I think we should first understand why it was done this way (use `git blame` to find PR that introduced it) and if it is ok to change it.
Also, it should be done in a dedicated commit explaining all this

hakehuang replied:
it is Peter Bigot add this original with  PR #73638 with commit #ae7a59eae4b877b1990c576b307f185c3a369484
```
test/drivers/gpio: add tests for new API

Test that the new port API functions all behave as expected, including
physical vs logical level for input and output as well as masked and
set-based output operations.  Also tests the new pin API functions.

For running on real hardware this test now uses a local test-specific
devicetree binding.  For build-only tests any platform with a GPIO
alias should be tested.

The new code increases flash requirements so add a filter to exclude
platforms that won't link.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
```

but there is not related comments

hakehuang replied:
> Also, it should be done in a dedicated commit explaining all this

updated

hakehuang committed:
boards: lpcxpresso55s36 : add arduino_gpio

add arduino_gpio in support list

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
tests: gpio: gpio_basic_api: update customerized board list

update boards that can not use the common pins
mimxrt102x reuse with ethernet
mimxrt106x has external pull up
lpcxpresso55s3x reuse with int

add CONFIG_SKIP_PULL_TEST by default to be more general
user can enable this test in customized case with select pin

ST prefer to using D9 D10 pair to testing.

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
tests: gpio: gpio_basic_api split input and ouput

input and output shall be split and configure by dts

fixes: #78381

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
boards: lpcxpresso55s28: add arduino_gpio support

add arduino_apio support to lpcxpresso55s28

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
boards: lpcxpresso55s69: add arduino_gpio support

add arduino_gpio support to lpcxpresso55s69

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang commented:
@erwango i enable the CONFIG_SKIP_PULL_TEST by default, which can save some efforts on select a gpio when the board has external pull up/down

hakehuang committed:
boards: da14695_dk_usb: remove the arduino_gpio support

da14695_dk_usb use mikro-bus interface not arduino

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
test: gpio: gpio_basic_api: enhance the gpio test

1. add delay read config for platforms that has capacitor with pin
2. add dts pin flags to gpio_pin_config when output.

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang commented:
all NXP platforms with arduino header interfaces platform test PASS.

hakehuang commented:
@erwango , is there any more issues found on ST platforms?

hakehuang commented:
@erwango how is your test status? all NXP platform PASS.

erwango commented:
> @erwango how is your test status? all NXP platform PASS.

Still on going, but I think it should be fine soon.

hakehuang commented:
> For everything to work well on the STM32 side, you should rather use pins D9 and D10, so Arduino headers 15 and 16.

@erwango I update ST overlay to use D9/D10, please check. Thanks

hakehuang commented:
> There were some communication issues with other tests on some series using pins D0 and D1 and a slight modification would be needed on the st_arduino.overlay.

As to this, you can check the whether you have PULL_UP on those pins, if so you need add a config to skip the PULL_TEST.

djiatsaf-st commented:
> > For everything to work well on the STM32 side, you should rather use pins D9 and D10, so Arduino headers 15 and 16.
> 
> @erwango I update ST overlay to use D9/D10, please check. Thanks

I have performed hw tests on this PR and it seems correct with D9/D10

hakehuang commented:
@erwango no the test pass on all NXP and ST platforms.

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -36,15 +36,27 @@ tests:
+      - mimxrt1060_evkb
+      - lpcxpresso55s69/lpc55s69/cpu0
+      - lpcxpresso55s36
+      - nucleo_wb05kz
+      - nucleo_wb09ke
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
Not required as filter is !CONFIG_SOC_FAMILY_STM32

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -53,3 +65,30 @@ tests:
+      - lpcxpresso55s36
+      - nucleo_wb09ke
+      - nucleo_wb05kz
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
From what I understood, only wb0 boards required a specific overlay. @djiatsaf-st can you clarify the status ?

hakehuang replied:
this is not a must, as it also helps to demo how to customize the test settings. agree? @erwango

djiatsaf-st replied:
> From what I understood, only wb0 boards required a specific overlay. @djiatsaf-st can you clarify the status ?

There will be no problem for nucleo_l031k6.
For nucleo_wba55cg, the user manual specifies that we need to change state of solder bridges to use pin D8 or D9. 
So it will be better to leave as excluded.

erwango replied:
@hakehuang Please make `nucleo_l031k6` use st_arduino.overlay  and we'll be good to go

hakehuang replied:
done

erwango commented on tests/drivers/gpio/gpio_basic_api/testcase.yaml:
```diff
@@ -53,3 +65,30 @@ tests:
+      # below boards need customized
+      - nucleo_wb05kz
+      - nucleo_wb09ke
+      - nucleo_wba55cg
+      - nucleo_l031k6
```
See upper comment

erwango commented on tests/drivers/gpio/gpio_basic_api/st_arduino.overlay:
```diff
@@ -4,13 +4,13 @@
-		out-gpios = <&arduino_header 6 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
-		in-gpios = <&arduino_header 7 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
+		out-gpios = <&arduino_header 15 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
+		in-gpios = <&arduino_header 16 (GPIO_ACTIVE_HIGH | GPIO_PUSH_PULL | GPIO_PULL_UP)>;
```
No need for a specific commit. This should be fixed directly in the commit that introduced the file.

hakehuang replied:
ok

hakehuang commented:
the build failure is not related to this PR

hakehuang commented:
the ci failure is not related to this PR, but #79576

erwango commented on tests/drivers/gpio/gpio_basic_api/st_arduino.overlay:
Suggestion:
I expect non STM32 boards can also use this overlay for specific constraints.
So I suggest to rename:
arduino.overlay > arduino_D0D1.overlay
arduino_st.overlay > arduino_D9D10.overlay

which will make it more generic and reuseable by others

erwango reviewed (changes requested):
Last 2 comments

erwango reviewed (approved):
Thanks for the changes

erwango commented:
@nashif Please have a look
